resources/src/effect/mod.rs

```diff
@@ -0,0 +1,562 @@
+use amount::Amount;
+use asset::{AssetIdentifier, Flag};
+use serde::{de, Deserialize, Deserializer};
+mod account_created;
+mod account_credited;
+mod account_debited;
+mod account_flags_updated;
+mod account_home_domain_updated;
+mod account_removed;
+mod account_thresholds_updated;
+mod signer_created;
+mod signer_removed;
+mod signer_updated;
+mod trade;
+mod trustline_authorized;
+mod trustline_created;
+mod trustline_deauthorized;
+mod trustline_removed;
+mod trustline_updated;
+
+pub use self::account_created::AccountCreated;
+pub use self::account_credited::AccountCredited;
+pub use self::account_debited::AccountDebited;
+pub use self::account_flags_updated::AccountFlagsUpdated;
+pub use self::account_home_domain_updated::AccountHomeDomainUpdated;
+pub use self::account_removed::AccountRemoved;
+pub use self::account_thresholds_updated::AccountThresholdsUpdated;
+pub use self::signer_created::SignerCreated;
+pub use self::signer_removed::SignerRemoved;
+pub use self::signer_updated::SignerUpdated;
+pub use self::trade::Trade;
+pub use self::trustline_authorized::TrustlineAuthorized;
+pub use self::trustline_created::TrustlineCreated;
+pub use self::trustline_deauthorized::TrustlineDeauthorized;
+pub use self::trustline_removed::TrustlineRemoved;
+pub use self::trustline_updated::TrustlineUpdated;
+
+#[cfg(test)]
+mod test;
+
+/// A successful operation will yield zero or more effects. These effects represent specific
+/// changes that occur in the ledger, but are not necessarily directly reflected in the ledger or
+/// history, as transactions and operations are.
+#[derive(Debug)]
+pub struct Effect {
+    id: String,
+    paging_token: String,
+    kind: Kind,
+}
+
+/// Each effect type is representing by a kind and captures data specific to that
+/// type within it's newtype.
+#[derive(Debug, Deserialize)]
+pub enum EffectKind {
+    /// An effect representing the fact that an account was created
+    AccountCreated(AccountCreated),
```

Comment:
Going along with the idea of modules per type of effect, you might have an enum for that as well. It becomes more nested and that might be bad as well, but I could see something like:

```rust
enum Kind {
  Account(account::Kind)
  ..
}
```

and then in the account module:
```rust
enum Kind {
  Created(AccountCreated),
  Removed(AccountRemoved),
}
```

Not sure that helps, exactly, but it does allow for potentially easy checking what the affect applies to:

```rust
match effect.kind() {
  Kind::Account(_) => {},
  Kind::Signer(_) => ...
}
```